### PR TITLE
Promise usage

### PR DESCRIPTION
**Pull request by realdennis**
Refer to the issue https://github.com/yahoo/ycb-config/issues/21 I said, if we could return promise in our async function, it will reduce our function nested callback, I use the new method name `readPromise` which just a wrap of orignal method `read` and pass the callback to exchange fulfilled / rejected state as `promisify` do.

The testing code are just a POC for all `read()` functionality, we could make promised interface easily , actually we only need to test the correct `thisArgs` are passed to the original method in the future.

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by redonkulus** [comment]
Could just use util.promisify from nodejs to convert the callback apis too. Should also update the ycb lib for parity.

**Comment by realdennis** [comment]
Update for our asynchronous methods, and also test for :
- Config.prototype.readPromise
- Config.prototype.readNoMergePromise
- Config.prototype.readDimensionsPromise

**Comment by realdennis** [comment]
@redonkulus Looks like we use lib ycb synchronously, and I found the only async and public method exposed from lib ycb is `walksetting`, I'll update it later, if these two are dependency, please do not merge now.

**Comment by tom76kimo** [comment]
Hmm..what if we use `ycb-config` on client side? `util` won't work.

**Comment by redonkulus** [comment]
Generally this library isn't used on the client side and has always assumed server since it uses `fs` and other node specific libraries.

File: tests/lib/index.js [hunk 1]
```diff
@@ -1063,5 +1063,481 @@ describe('config', function () {
...
+            var config = new Config();
+            config.readPromise("foo", "bar", {}).catch(err => {
+              expect(err.message).to.equal('Unknown bundle "foo"');
+              next();
+            });
+          });
+
+          it("fails on unknown config", function(next) {
+            var config = new Config();
+            config.addConfig(
```
Comment (redonkulus) [inline tests/lib/index.js]: can these tests use `config.addConfigPromise` to avoid the callback pyramid?

**Comment by tom76kimo** [comment]
@redonkulus Oh, saw it. Thanks.

File: tests/lib/index.js (hunk 1 above)
Comment (realdennis) [reply tests/lib/index.js]: Updated.

File: tests/lib/index.js [hunk 2]
```diff
@@ -1063,5 +1063,445 @@ describe('config', function () {
...
+          it("reads contextualized .json config files", function(next) {
+            var config = new Config();
+            config
+              .addConfigPromise(
+                "modown",
+                "dimensions",
+                libpath.resolve(mojito, "node_modules/modown/dimensions.json")
+              )
+              .then(() =>
+                config
```
Comment (realdennis) [inline tests/lib/index.js]: Actually, I intentionally return promise directly, we could align the `then` or `catch` outside in the same level if you want.

Comment (redonkulus) [reply tests/lib/index.js]: Ya that would be better then  the  nested  promise  handling and allow us to use async await more easily  when we drop node 8 support.

**Comment by redonkulus** [comment]
@realdennis FYI I haven't merged  this yet because I'm going to re-release #20 in v2 of this package. Then we can add your changes to that.

File: tests/lib/index.js (hunk 2 above)
Comment (realdennis) [reply tests/lib/index.js]: Agreed.

**Comment by redonkulus** [comment]
Can you add documentation to the readme about the promise api option?

**Comment by realdennis** [comment]
@redonkulus 

Before adding the Promise API documentation, I have an idea and would like to discuss it with you.

I create a new branch to design another way that calls promises method.

```javascript
config.promises.read('simple', 'routes', {}))
```
There are also some examples of `config.promises.XXX` in the test code of the new branch.
https://github.com/realdennis/ycb-config/blob/promises/tests/lib/index.js#L1078-L1094

It'll be **more similar to node built-in modules**, such as `fs.promises.readFile`.

---

But there is a limitation, we can't get and bind the instance "this" in the nested prototype method (such as `Config.prototype.promises.read`), so this is why I define all methods with "promise" suffix first.

Now, there is a quick solution in branch function code, just define "promises" object in the constructor.
https://github.com/realdennis/ycb-config/blob/promises/lib/index.js#L196-L202

If you think this interface would be better and have any suggestions for this branch, please let me know and I will create a new pull request for discussion, otherwise I will continue to add the document for the Promise suffix usage.

**Comment by redonkulus** [comment]
Funny you mention this. I was going to suggest this type of api myself. I’m fine with defining it in the constructor to bind it properly.

**Comment by redonkulus** [comment]
Closed in favor of #24

**Pull request closed without merging**